core/src/main/java/com/linecorp/armeria/common/RequestContextAwareCompletableFuture.java

```diff
@@ -243,4 +253,96 @@
+        final Callable<? extends T> callable = Functions.toCallable(supplier);
+        return (CompletableFuture<T>) ctx.makeContextAware(supplyAsync(
+                Functions.fromCallable(ctx.makeContextAware(callable))));
+    }
+
+    @SuppressWarnings("unchecked")
+    public CompletableFuture<T> completeAsync(Supplier<? extends T> supplier,
+                                              Executor executor) {
+        final Callable<? extends T> callable = Functions.toCallable(supplier);
+        return (CompletableFuture<T>) ctx.makeContextAware(supplyAsync
+                (Functions.fromCallable(ctx.makeContextAware(callable)), executor));
+    }
+
+    public CompletableFuture<T> orTimeout(long timeout, TimeUnit unit) {
+        if (unit == null)
+            throw new NullPointerException();
+        if (!this.isDone())
+            whenComplete(new Canceller(Delayer.delay(new Timeout(this),
+                    timeout, unit)));
+        return ctx.makeContextAware(this);
+    }
+
+    public CompletableFuture<T> completeOnTimeout(T value, long timeout,
+                                                  TimeUnit unit) {
+        if (unit == null)
+            throw new NullPointerException();
+        if (!this.isDone())
+            whenComplete(new Canceller(Delayer.delay(
+                    new DelayedCompleter<>(this, value),
+                    timeout, unit)));
+        return ctx.makeContextAware(this);
+    }
+
+    static final class Delayer {
+        static ScheduledFuture<?> delay(Runnable command, long delay,
+                                        TimeUnit unit) {
+            return delayer.schedule(command, delay, unit);
+        }
+
+        static final class DaemonThreadFactory implements ThreadFactory {
+            public Thread newThread(Runnable r) {
+                Thread t = new Thread(r);
+                t.setDaemon(true);
+                t.setName("CompletableFutureDelayScheduler");
+                return t;
+            }
+        }
+
+        static final ScheduledThreadPoolExecutor delayer;
+        static {
+            (delayer = new ScheduledThreadPoolExecutor(
+                    1, new DaemonThreadFactory())).
+                    setRemoveOnCancelPolicy(true);
+        }
+    }
+
+    static final class Timeout implements Runnable {
+        final CompletableFuture<?> f;
+        Timeout(CompletableFuture<?> f) { this.f = f; }
+        public void run() {
+            if (f != null && !f.isDone())
+                f.completeExceptionally(new TimeoutException());
+        }
+    }
+
+    static final class DelayedCompleter<U> implements Runnable {
+        final CompletableFuture<U> f;
+        final U u;
+        DelayedCompleter(CompletableFuture<U> f, U u) { this.f = f; this.u = u; }
+        public void run() {
+            if (f != null)
+                f.complete(u);
+        }
+    }
+
+    static final class Canceller implements BiConsumer<Object, Throwable> {
```

Comment:
Could implement `BiFunction` and return `null` to be used with `handle()`.

Reply:
Thank you for the good information 👍